python/cudf/cudf/core/buffer/utils.py

```diff
@@ -0,0 +1,218 @@
+    from cudf.core.column_accessor import ColumnAccessor
+    from cudf.core.frame import Frame
+    from cudf.core.indexed_frame import IndexedFrame
+
+    """Return all columns in `obj` (no duplicates)"""
+    found: List[Column] = []
+    found_ids: Set[int] = set()
+
+    def _get_columns(obj: object) -> None:
+        if isinstance(obj, Column):
+            if id(obj) not in found_ids:
+                found_ids.add(id(obj))
+                found.append(obj)
+        elif isinstance(obj, IndexedFrame):
+            _get_columns(obj._data)
+            _get_columns(obj._index)
+        elif isinstance(obj, Frame):
+            _get_columns(obj._data)
+        elif isinstance(obj, ColumnAccessor):
+            for o in obj.columns:
+                _get_columns(o)
+        elif isinstance(obj, (list, tuple)):
+            for o in obj:
+                _get_columns(o)
+        elif isinstance(obj, Mapping):
+            for o in obj.values():
+                _get_columns(o)
+
+    _get_columns(obj)
+    return found
+
+
+def mark_columns_as_read_only_inplace(obj: Any) -> None:
+    """
+    Mark all columns found in `obj` as read-only.
+
+    This is an in-place operation, which does nothing if
+    spilling is disabled.
+
+    Making columns as ready-only, makes it possible to unspill the
+    underlying buffers partially.
+    """
+
+    if global_manager_get() is None:
+        return
+
+    for col in get_columns(obj):
+        if col.base_children:
+            continue  # TODO: support non-fixed-length data types
+
+        if col.base_mask is not None:
+            continue  # TODO: support masks
+
+        if col.base_data is None:
+            continue
+        assert col.data is not None
+
+        if col.data is col.base_data:
+            continue  # We can ignore non-views
+
+        if isinstance(col.base_data, SpillableBuffer) and isinstance(
+            col.data, SpillableBuffer
+        ):
+            with col.base_data.lock:
+                if not col.base_data.is_spilled:
+                    continue  # We can ignore non-spilled columns
+                mem = col.data.memoryview()
+            col.set_base_data(as_device_buffer_like(mem, exposed=False))
+            col._offset = 0
+
+
+_thread_spill_locks: Dict[int, Tuple[Optional[SpillLock], int]] = {}
+
+
+def get_spill_lock():
+    _id = threading.get_ident()
+    spill_lock, _ = _thread_spill_locks.get(_id, (None, 0))
+    return spill_lock
+
+
+def push_thread_spill_lock():
+    _id = threading.get_ident()
+    spill_lock, count = _thread_spill_locks.get(_id, (None, 0))
+    if spill_lock is None:
+        spill_lock = SpillLock()
+    _thread_spill_locks[_id] = (spill_lock, count + 1)
+
+
+def pop_thread_spill_lock():
+    _id = threading.get_ident()
+    spill_lock, count = _thread_spill_locks[_id]
+    if count == 1:
+        spill_lock = None
+    _thread_spill_locks[_id] = (spill_lock, count - 1)
+
+
+def with_spill_lock(*, read_only_columns=False):
```

Comment:
I did a micro benchmark on my local workstation. 
The overhead is ~0.2us when spilling is disabled and ~0.7us when enabled. 
When `read_only_columns=True` the overhead is `127us`. 

See https://gist.github.com/madsbk/da6520e7583cf5d728a1b5a1b09200f3

Reply:
OK so `read_only_columns` is what introduces significant slowdown, otherwise it's pretty low overhead. I am OK with that since that's in the spilling code path anyway. As long as we don't slow down code that does not do any spilling just by adding the context manager, I don't mind if there is some overhead in the spilling logic.

Reply:
Agree